pkg/common/health/cache.go

```diff
@@ -0,0 +1,192 @@
+	// failing health check.
+	details State
+
+	// checkTime is the time of the last health check
+	checkTime time.Time
+
+	// contiguousFailures the number of failures that occurred in a row
+	contiguousFailures int64
+
+	// timeOfFirstFailure the time of the initial transitional failure for
+	// any given health check
+	timeOfFirstFailure time.Time
+}
+
+func newCache(log logrus.FieldLogger, clock clock.Clock) *cache {
+	return &cache{
+		allowedChecks: make(map[string]Checkable),
+		currentStatus: make(map[string]checkState),
+		log:           log,
+		clk:           clock,
+	}
+}
+
+type cache struct {
+	allowedChecks map[string]Checkable
+
+	currentStatus map[string]checkState
+	mtx           sync.RWMutex
+	clk           clock.Clock
+
+	log   logrus.FieldLogger
+	hooks struct {
+		statusUpdated chan struct{}
+	}
+}
+
+func (c *cache) addCheck(name string, checkable Checkable) error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if _, ok := c.allowedChecks[name]; ok {
+		return fmt.Errorf("check %q has already been added", name)
+	}
+
+	c.allowedChecks[name] = checkable
+	return nil
+}
+
+func (c *cache) getStatuses() map[string]checkState {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	statuses := make(map[string]checkState, len(c.currentStatus))
+	for k, v := range c.currentStatus {
+		statuses[k] = v
+	}
+
+	return statuses
+}
+
+func (c *cache) start(ctx context.Context) error {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	if len(c.allowedChecks) < 1 {
+		return errors.New("no health checks defined")
+	}
+
+	c.startRunner(ctx)
+	return nil
+}
+
+func (c *cache) startRunner(ctx context.Context) {
+	c.log.Debug("Initializing health checkers")
+	checkFunc := func() {
+		for name, check := range c.allowedChecks {
+			state, err := verifyStatus(check)
+
+			checkState := checkState{
+				details:   state,
+				checkTime: c.clk.Now(),
+			}
+			if err != nil {
+				c.log.WithField("check", name).
+					WithError(err).
+					Error("Health check has failed")
+				checkState.err = err
+			}
+
+			c.setStatus(name, checkState)
+		}
+		if c.hooks.statusUpdated != nil {
+			c.hooks.statusUpdated <- struct{}{}
+		}
+	}
+
+	ticker := c.clk.Ticker(readyCheckInterval)
+
+	go func() {
+		defer func() {
+			c.log.Debug("Finishing health checker")
+			ticker.Stop()
+		}()
+		for {
+			checkFunc()
+
+			select {
+			case <-ticker.C:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (c *cache) setStatus(name string, state checkState) {
+	c.embellishState(name, &state)
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	c.currentStatus[name] = state
+}
+
+func (c *cache) embellishState(name string, state *checkState) {
+	// get the previous state
+	c.mtx.RLock()
+	prevState := c.currentStatus[name]
+	c.mtx.RUnlock()
+
+	switch {
+	case state.err != nil && prevState.err == nil:
+		// State start to fail, add log and set failures tracking
+		c.log.WithFields(logrus.Fields{
+			"check":   name,
+			"details": state.details,
+			"error":   state.err.Error(),
```

Comment:
for better or worse, we usually put these field names in the common/telemetry package